src/runtime/runtime_init.c

```diff
@@ -50,11 +55,23 @@ static void format_number(buffer dest, struct formatter_state *s, vlist *a)
     }
     if (s->format == 'd' && x < 0) {
 	/* emit sign & two's complement */
-        push_u8(dest, '-');
+        sign = 1;
         x = -x;
     }
-
-    print_number(dest, x, base, s->width);
+    print_number(tmp, x, base, s->precision);
+    int len = buffer_length(tmp) + sign;
+    if (s->precision == 0 && x == 0)
+        len = 0;
+    if (sign && s->fill == '0')
+        push_u8(dest, '-');
+    if (len < s->width && s->align == 0)
+        for (int i = 0; i < s->width - len; i++) push_u8(dest, s->fill);
```

Comment:
Nit: duplicate for loops could be made into 'fill' helper function.

Reply:
Good point, I've cleaned that up with a helper function.